File: 백준_C++/4179.cpp
#include <bits/stdc++.h>
using namespace std;

int R, C, fire[1001][1001], dist[1001][1001];
string board[1001];
int dx[4] = {0, 0, -1, 1};
int dy[4] = {1, -1, 0, 0};
int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> R >> C;
    for (int i = 0; i < R; i++) cin >> board[i];
    queue<pair<int, int>> Q;
    pair<int, int> start;
    for (int i = 0; i < R; i++) {
        for (int j = 0; j < C; j++) {
            if (board[i][j] == 'J') start = {i, j};
            if (board[i][j] == 'F') {
                Q.push({i, j});
                fire[i][j] = 1;
            }
        }
    }

    while(!Q.empty()) {
        auto cur = Q.front(); Q.pop();
        for (int i = 0; i < 4; i++) {
            int nx = cur.first + dx[i];
            int ny = cur.second + dy[i];
            if (nx < 0 || nx >= R || ny < 0 || ny >= C) continue;
            if (fire[nx][ny] > 0 || board[nx][ny] == '#') continue;
            fire[nx][ny] = fire[cur.first][cur.second] + 1;
            Q.push({nx, ny});
        }
    }

    Q.push({start});
    dist[start.first][start.second] = 1;
    while(!Q.empty()) {
        auto cur = Q.front(); Q.pop();
        for (int i = 0; i < 4; i++) {
            int nx = cur.first + dx[i];
            int ny = cur.second + dy[i];
            if (nx == -1 || nx == R || ny == -1 || ny == C) {
                cout << dist[cur.first][cur.second];
                return 0;
            } 
            if (dist[nx][ny] != 0 || board[nx][ny] == '#') continue;
            if (fire[nx][ny] != 0 && dist[cur.first][cur.second] + 1 >= fire[nx][ny]) continue;

            dist[nx][ny] = dist[cur.first][cur.second] + 1;
            Q.push({nx, ny});
        }
    }
    cout << "IMPOSSIBLE";
}

File: 백준_Java/4179.java
import java.io.*;
import java.util.*;

class Main {
    static final int[] dx = {0, 0, 1, -1};
    static final int[] dy = {1, -1, 0, 0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        char[][] board = new char[n][m];

        for (int i = 0; i < n; i++) {
            String lines = br.readLine();
            for (int j = 0; j < m; j++) {
                board[i][j] = lines.charAt(j);
            }
        }

        int result = solution(n, m, board);

        if (result == -1) {
            System.out.print("IMPOSSIBLE");
        } else {
            System.out.print(result);
        }
    }

    static int solution(int n, int m, char[][] board) {
        int[][] fire = new int[n][m];
        int[][] dist = new int[n][m];
        Queue<Pair> q = new LinkedList<Pair>();
        Pair start = null;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == 'F') {
                    q.add(new Pair(i, j));
                    fire[i][j] = 1;
                }

                if (board[i][j] == 'J') {
                    start = new Pair(i, j);
                    dist[i][j] = 1;
                }
            }
        }

        while(!q.isEmpty()) {
            Pair cur = q.poll();
            for (int dir = 0; dir < 4; dir++) {
                int nx = cur.x + dx[dir];
                int ny = cur.y + dy[dir];

                if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if (board[nx][ny] == '#' || fire[nx][ny] > 0) continue;
                fire[nx][ny] = fire[cur.x][cur.y] + 1;
                q.add(new Pair(nx, ny));
            }
        }

        q.add(start);
        while(!q.isEmpty()) {
            Pair cur = q.poll();
            for (int dir = 0; dir < 4; dir++) {
                int nx = cur.x + dx[dir];
                int ny = cur.y + dy[dir];

                if (nx < 0 || nx >= n || ny < 0 || ny >= m) {
                    return dist[cur.x][cur.y];
                }
                if (board[nx][ny] == '#' || dist[nx][ny] > 0) continue;
                if (fire[nx][ny] != 0 && fire[nx][ny] <= dist[cur.x][cur.y] + 1) continue;
                dist[nx][ny] = dist[cur.x][cur.y] + 1;
                q.add(new Pair(nx, ny));
            }
        }
        
        return -1;
    }

    static class Pair{
        int x;
        int y;

        public Pair(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}